keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,992 @@
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+#### Alpha
+
+- Feature disabled by default.
+- Unit and integration tests completed and passed.
+- Additional tests are in Testgrid and linked in KEP
+
+#### Beta
+
+- Feature enabled by default.
+- Gather feedback from developers and out-of-tree plugins.
+- Benchmark tests passed, and there is no performance degradation.
+- Update documents to reflect the changes.
+
+#### GA
+
+- Fix all reported bugs.
+- Feature enabled and cannot be disabled. All feature gate guarded logic are removed.
+- Update documents to reflect the changes.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+- Upgrade
+  - While the feature gate is enabled, `.spec.schedulingPaused` can be mutated by users.
+  - While the feature gate is enabled and `.spec.schedulingPaused` is not set, a default value will
+  be given to behave the same like it's specified by users.
+- Downgrade
+  - Previously configured values will be ignored.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+N/A. The feature doesn't affect nodes.
```

Comment:
True. But a situation where api-server is vCurrent+1 and scheduler is vCurrent things will be unpredictable specially for inexperienced users.

As we stand today i don't think we have a solution for this. Normally it has no impact on the behavior because old clients don't see new fields but this field is of high significance (high expectations when set to `true`). 

I think we will have to clarify the documentation specifically covering scheduler v api-server version skew for upgrade and downgrade.

Reply:
I would go even further:

> New fields must be additive.  The new paused field, due to forward compatibility requirements, may be ignored by custom schedulers for long periods.  Our default scheduler will ignore it unless (configuration / gates / version).

"paused" isn't special, schedulers may ignore it, and we can't break custom schedulers through its introduction.

We *should* do what we can to make administrators aware that the feature is not implemented or enabled - that might be something like a metric on "pods scheduled without clearing schedulingPaused explicitly", as well as warnings (possibly) reported to clients that do the scheduling without clearing that value.  But it will always be supported to set `spec.nodeName` without clearing `schedulingPaused`, and it should be possible as you note to discover whether a component is respecting paused.

Reply:
> we can't break custom schedulers through its introduction

@ahg-g I think this is the hard requirement of v1 guarantee. In other words, if API Server is at vCurrent and the feature is enabled, and scheduler at vCurrent-1, and it happens that some controller sets the paused flag, we need to ensure this pod get scheduled properly. (The scheduler can be custom scheduler or admin). So, like Clayton mentioned, it's a must to:

> But it will always be supported to set spec.nodeName without clearing schedulingPaused

So our discussion on erroring out "setting nodeName on a paused pod" is not an option at all.

Reply:
> We should do what we can to make administrators aware that the feature is not implemented or enabled

Metric works in this aspect. Default scheduler can generate some logs when it detects some scheduled pod carries the paused field. Other than that, it seems there is no good way to "warn" the admin/user. Maybe we can dump a warning in kubectl's output to alert on scheduled Pod carrying paused state?

Reply:
Reading this again this morning, I want to nuance my statement and define what constitutes "break". I think @ahg-g makes good points in the other thread, and I'm trying to find the right use case to guide what constitutes a breakage (I may be thinking too strictly):

TL:DR: I think blocking `spec.nodeName` when `schedulingPaused` is safe because the use of schedulingPaused is opt-in for users - and we can't set schedulingPaused automatically until the feature is in use for 1 release.

If a user (human or workload controller) creates a pod without setting scheduling paused (regardless of whether this feature is on), they expect the pod to be scheduled as long as it fits on a node.  An administrator may override the default user-observable behavior of scheduling by: adding admission hooks that assign a non-default scheduler, adding admission hooks that alter the object to make it not fit, or customizing the default scheduler to have different behaviors.  These are all administrator defined variants that are allowed deviations.  A user expects scheduler to fit or not quickly, but there is no hard time boundary to schedule.  Users cannot observe the inner workings of the scheduler except as the scheduler chooses via status conditions.  Ultimately the only breakage a user will notice is a pod that sits forever in schedulable without some reason being given via the Schedulable condition - so we have wide latitude to change the scheduler flow by default.

We expose to a scheduler (default or custom) a similar contract - they observe a pod via the API in the `Schedulable` lifecycle state and are expected to decide which scheduler has responsibility based on a) default configuration and schedulerName, and b) administrative configuration.  Schedulers may make arbitrary decisions on scheduling, including using spec / metadata fields (or schedulerName) to push pods through workflows (i.e. schedulerName -> `custom/measure` -> `custom/wait_for_reservation`).  Such a scheduler must not be broken by the addition of a new lifecycle state, and in general custom schedulers should not break *if the feature is not leveraged*.

The addition of fields that alter the behavior of the scheduler for pods in the `Schedulable` lifecycle state is not new - for instance topology constraints, tolerations, and custom resources all required the user or admin to opt-in to leveraging those features through explicit action (by setting the field, configuring taints on nodes, or adding admission hooks that defaulted values).  Historically we have said the scheduler simply ignores those fields until updated, and then begins enforcing them (e.g. during an upgrade from alpha to beta, where the gate becomes available).  This is generally understood to be "not common enough for us to care about" :)

The behavior about blocking `spec.nodeName` being set if `schedulingPaused` is effectively `true` is an added capability above this - the user *has* opted in (and `schedulingPaused` is either user set or admin set, just like any other constraint), but we can enforce the logic at the apiserver level.  The scheduler that isn't aware of `schedulingPaused` and is now prevented from scheduling is only blocked due to the user's action (setting schedulingPaused) vs the behavior of the upgrade.

What we cannot do is set schedulingPaused by default on all new pods during a Kube upgrade without admin or user intervention.  That *would* break those unknowing schedulers.  An admin is allowed to enforce that (perhaps via an opt-in config flag), but we cannot by default without changing the observable behavior of the pod `spec.nodeName` field.

Reply:
> The scheduler that isn't aware of schedulingPaused and is now prevented from scheduling is only blocked due to the user's action (setting schedulingPaused) vs the behavior of the upgrade.

I _thought_ this is part of v1 guarantee..

> What we cannot do is set schedulingPaused by default on all new pods during a Kube upgrade without admin or user intervention.

No, we won't.

> we can't set schedulingPaused automatically until the feature is in use for 1 release

Even the feature reaches beta/GA, we won't make pods paused by default - it will still be opt-in feature.

---

> Ultimately the only breakage a user will notice is a pod that sits forever in schedulable without some reason being given via the Schedulable condition

This may be something worth discussing (also mentioned a bit in https://github.com/kubernetes/enhancements/pull/3522#discussion_r978872725), _if_ we expose a new reason on an existing condition type (`PodScheduled`), as you mentioned, the clients would have context to understand what's going on here, I read it as a good-to-have; however, if we consider some integrators like CA, from their perspective, they just look at the condition `PodScheduled` to treat it as unschedulable, and try to spin up a new instance. So with this regard, this will break their workflow.

Reply:
Indeed, this is an opt-in feature and will never be defaulted to "not ready" unless through a webhook, which is admin controlled. This feature shouldn't really be used in a cluster with a custom scheduler that doesn't support it yet, and I think it is appropriate for the api-server to be strict on that by rejecting setting nodeName when the pod is not ready.

As you mentioned, ideally all scheduling constraints should be this way, but unfortunately we can't enforce them via the api-server without introducing significant overhead.

Reply:
Sounds like we're all in agreement then, we should fail the attempt to set `spec.nodeName` while paused with a 409 conflict / validation error as originally proposed?

Reply:
Yes, a 409 conflict / validation error should return in this case.

Reply:
updated.